pallets/auctions/src/lib.rs

```diff
@@ -160,171 +154,297 @@ pub mod pallet {
+		ensure!(
+			token_owner == Some(general_data.owner.clone()),
+			Error::<T>::NotATokenOwner
+		);
+		ensure!(!&general_data.closed, Error::<T>::CannotSetAuctionClosed);
+
 		Ok(())
 	}
-}
 
-impl<T: Config> Auction<T::AccountId, T::BlockNumber, T::ClassId, T::InstanceId> for Pallet<T> {
-	type AuctionId = T::AuctionId;
-	type Balance = BalanceOf<T>;
-	type AccountId = T::AccountId;
+	fn validate_create(general_data: &GeneralAuctionData<T>) -> DispatchResult {
+		let is_transferrable = pallet_uniques::Pallet::<T>::can_transfer(&general_data.token.0, &general_data.token.1);
+		ensure!(is_transferrable, Error::<T>::TokenFrozen);
 
-	fn new_auction(info: AuctionInfoOf<T>) -> result::Result<Self::AuctionId, DispatchError> {
-		// Basic checks before an auction is created
-		Self::check_new_auction(&info)?;
-		let auction_id = <NextAuctionId<T>>::try_mutate(|next_id| -> result::Result<Self::AuctionId, DispatchError> {
+		Ok(())
+	}
+
+	fn handle_create(
+		sender: <T>::AccountId,
+		auction: &Auction<T>,
+		general_data: &GeneralAuctionData<T>,
+	) -> DispatchResult {
+		let auction_id = <NextAuctionId<T>>::try_mutate(|next_id| -> result::Result<<T>::AuctionId, DispatchError> {
 			let current_id = *next_id;
 			*next_id = next_id
 				.checked_add(&One::one())
 				.ok_or(Error::<T>::NoAvailableAuctionId)?;
 			Ok(current_id)
 		})?;
 
-		<Auctions<T>>::insert(auction_id, info.clone());
-		<AuctionOwnerById<T>>::insert(auction_id, &info.owner);
-		<AuctionEndTime<T>>::insert(info.end, auction_id, ());
+		<Auctions<T>>::insert(auction_id, auction.clone());
+		<AuctionOwnerById<T>>::insert(auction_id, &sender);
 
-		pallet_uniques::Pallet::<T>::freeze(RawOrigin::Signed(info.owner.clone()).into(), info.token.0, info.token.1)?;
+		pallet_uniques::Pallet::<T>::freeze(
+			RawOrigin::Signed(sender.clone()).into(),
+			general_data.token.0,
+			general_data.token.1,
+		)?;
 
-		Ok(auction_id)
+		Self::deposit_event(Event::AuctionCreated(sender, auction_id));
+
+		Ok(())
 	}
 
-	fn update_auction(id: Self::AuctionId, info: AuctionInfoOf<T>) -> DispatchResult {
-		<Auctions<T>>::try_mutate(id, |auction| -> DispatchResult {
-			ensure!(auction.is_some(), Error::<T>::AuctionNotExist);
-			*auction = Some(info);
-			Ok(())
+	fn validate_update(sender: <T>::AccountId, general_data: &GeneralAuctionData<T>) -> DispatchResult {
+		ensure!(general_data.owner == sender, Error::<T>::NotAuctionOwner);
+
+		let current_block_number = frame_system::Pallet::<T>::block_number();
+		ensure!(
+			current_block_number < general_data.start,
+			Error::<T>::AuctionAlreadyStarted
+		);
+
+		Ok(())
+	}
+
+	fn handle_update(
+		sender: <T>::AccountId,
+		auction_id: T::AuctionId,
+		updated_auction: Auction<T>,
+		general_data: &GeneralAuctionData<T>,
+	) -> DispatchResult {
+		<Auctions<T>>::try_mutate(auction_id, |auction_result| -> DispatchResult {
+			if let Some(_auction) = auction_result {
+				Self::validate_update(sender, &general_data)?;
```

Comment:
validate_update already takes a reference.

```suggestion
				Self::validate_update(sender, general_data)?;
```